app/discord/activityTracker.ts

```diff
@@ -0,0 +1,95 @@
+import { Events, ChannelType } from "discord.js";
+import type { Client, Message, PartialMessage, TextChannel } from "discord.js";
+import db from "~/db.server";
+
+export async function startActivityTracking(client: Client) {
+  const channelCache = new Map<string, TextChannel>();
+
+  async function getOrFetchChannel(msg: Message) {
+    // TODO: cache eviction?
+    return channelCache.has(msg.channelId)
+      ? channelCache.get(msg.channelId)
+      : channelCache
+          .set(msg.channelId, (await msg.channel.fetch()) as TextChannel)
+          .get(msg.channelId);
+  }
+
+  client.on(Events.MessageCreate, async (msg) => {
+    const info = await getMessageStats(msg);
+    if (!info) return;
+    await db
+      .insertInto("message_stats")
+      .values({
+        ...info,
+        message_id: msg.id,
+        author_id: msg.author!.id,
+        guild_id: msg.guildId!,
+        channel_id: msg.channelId,
+        recipient_id: msg.mentions?.repliedUser?.id ?? null,
+        channel_category: (await getOrFetchChannel(msg))?.parent?.name,
+      })
+      .execute();
+  });
+
+  client.on(Events.MessageUpdate, async (msg) => {
+    const info = await getMessageStats(msg);
+    if (!info) return;
+    await updateStatsById(msg.id).set(info).execute();
+  });
+
+  client.on(Events.MessageDelete, async (msg) => {
+    const info = await getMessageStats(msg);
+    if (!info) return;
+    await db
+      .deleteFrom("message_stats")
+      .where("message_id", "=", msg.id)
+      .execute();
+  });
+
+  client.on(Events.MessageReactionAdd, async (msg) => {
+    await updateStatsById(msg.message.id)
+      .set({ react_count: (eb) => eb(eb.ref("react_count"), "+", 1) })
+      .execute();
+  });
+
+  client.on(Events.MessageReactionRemove, async (msg) => {
+    await updateStatsById(msg.message.id)
+      .set({ react_count: (eb) => eb(eb.ref("react_count"), "-", 1) })
+      .execute();
+  });
+}
+
+function updateStatsById(id: string) {
+  return db.updateTable("message_stats").where("message_id", "=", id);
+}
+
+async function getMessageStats(msg: Message | PartialMessage) {
+  // TODO: more filters
+  if (msg.channel.type !== ChannelType.GuildText || msg.author?.bot) {
+    return;
+  }
+  return {
+    char_count: msg.content?.length ?? 0,
+    word_count: msg.content?.split(/\s+/).length ?? 0,
+    react_count: msg.reactions.cache.size,
+    sent_at: String(msg.createdTimestamp),
```

Comment:
Can this be a UTC datetime instead of unix time?

Reply:
it *can*, imo with sqlite it's easier to store as int (since [dates dont really exist in sqlite](https://www.sqlite.org/datatype3.html)) and then format to whatever you like when querying

Reply:
From chat:
> sqlite is basically "weakly typed" and doesn't have an actual date type, but it can interchangeably work with dates in text or int formats (also real but :grimacing:)
> sql standard supports a bunch of data types, but in sqlite they are all aliases for one of: int text blob real blob or null

Seems good to me 👍